py/desispec/preproc.py

```diff
@@ -0,0 +1,196 @@
+      * FITS upperlimit-inclusive vs. python upperlimit-exclusive
+      * FITS[x,y] vs. python[y,x]
+    
+    i.e. BIASSECB='[7:56,51:4146]' -> (slice(50,4146), slice(6,56))
+    '''
+    m = re.search('\[(\d+):(\d+)\,(\d+):(\d+)\]', value)
+    if m is None:
+        raise ValueError, 'unable to parse {} as [a:b, c:d]'.format(value)
+    else:
+        xmin, xmax, ymin, ymax = map(int, m.groups())
+        
+    return np.s_[ymin-1:ymax, xmin-1:xmax]
+
+def _overscan(pix, sigma=5, niter=3):
+    '''
+    returns overscan, readnoise from overscan image pixels
+    
+    Optional:
+        sigma : sigma clipping parameter for mean calculation
+        niter : number of iterative refits
+    '''
+    #- normalized median absolute deviation as robust version of RMS
+    #- see https://en.wikipedia.org/wiki/Median_absolute_deviation
+    overscan = np.median(pix)
+    absdiff = np.abs(pix - overscan)
+    readnoise = 1.4826*np.median(absdiff)
+    
+    #- input pixels are integers, so iteratively refit
+    for i in range(niter):
+        absdiff = np.abs(pix - overscan)
+        good = absdiff < sigma*readnoise
+        overscan = np.mean(pix[good])
+        readnoise = np.std(pix[good])
+    
+    return overscan, readnoise
+
+def preproc(rawimage, header, bias=False, pixflat=False, mask=False):
+    '''
+    preprocess image using metadata in header
+
+    image = ((rawimage-bias-overscan)*gain)/pixflat
+    
+    Args:
+        rawimage : 2D numpy array directly from raw data file
+        header : dict-like metadata, e.g. from FITS header, with keywords
+            CAMERA, DATE-OBS, CCDSEC, BIASSECx, DATASECx, CCDSECx
+            where x = A, B, C, D for each of the 4 amplifiers
+        
+    Optional bias, pixflat, and mask can each be:
+        False: don't apply that step
+        True: use default calibration data for that night
+        ndarray: use that array
+        filename (str or unicode): read HDU 0 and use that
+
+    Returns Image object with member variables:
+        image : 2D preprocessed image in units of electrons per pixel
+        ivar : 2D inverse variance of image
+        mask : 2D mask of image (0=good)
+        readnoise : 2D per-pixel readnoise of image
+        meta : metadata dictionary
+            TODO: define what keywords are included
+
+    preprocessing includes the following steps:
+        - bias image subtraction
+        - overscan subtraction (from BIASSEC* keyword defined regions)
+        - readnoise estimation (from BIASSEC* keyword defined regions)
+        - gain correction (from GAIN* keywords)
+        - pixel flat correction
+        - cosmic ray masking
+        - propagation of input known bad pixel mask
+        - inverse variance estimation
+        
+    Notes:
+    
+    The bias image is subtracted before any other calculation to remove any
+    non-uniformities in the overscan regions prior to calculating overscan
+    levels and readnoise.
```

Comment:
Is this really the right thing to do? I've typically always modeled the overscan region (e.g., median average across rows or columns), subtracted from everything, including the bias frame, and then bias-subtracted.  Here you're taking the overscan level to be a scalar.